kuma/wiki/views/document.py

```diff
@@ -687,68 +673,118 @@ def document(request, document_slug, document_locale):
+    # The etag decorator not only adds the ETag header to the response,
+    # but also handles conditional responses based on the incoming and
+    # outgoing ETag headers.
+    @etag(get_etag)
+    def get_response(request, *args):
+        return _add_kuma_revision_header(doc, response)
 
-    # Retrieve pre-parsed content hunks
-    quick_links_html = doc.get_quick_links_html()
-    zone_subnav_html = doc.get_zone_subnav_html()
-    body_html = doc.get_body_html()
+    return get_response(request)
 
-    # Record the English slug in Google Analytics, to associate translations
-    if original_doc.locale == 'en-US':
-        en_slug = original_doc.slug
-    elif original_doc.parent and original_doc.parent.locale == 'en-US':
-        en_slug = original_doc.parent.slug
-    else:
-        en_slug = ''
-
-    share_text = ugettext(
-        'I learned about %(title)s on MDN.') % {"title": doc.title}
-
-    contributors = doc.contributors
-    contributors_count = len(contributors)
-    has_contributors = contributors_count > 0
-
-    # Bundle it all up and, finally, return.
-    context = {
-        'document': original_doc,
-        'document_html': doc_html,
-        'toc_html': toc_html,
-        'quick_links_html': quick_links_html,
-        'zone_subnav_html': zone_subnav_html,
-        'body_html': body_html,
-        'contributors': contributors,
-        'contributors_count': contributors_count,
-        'contributors_limit': 6,
-        'has_contributors': has_contributors,
-        'fallback_reason': fallback_reason,
-        'kumascript_errors': ks_errors,
-        'macro_sources': (kumascript.macro_sources(force_lowercase_keys=True)
-                          if ks_errors else
-                          None),
-        'render_raw_fallback': rendering_params['render_raw_fallback'],
-        'seo_summary': seo_summary,
-        'seo_parent_title': seo_parent_title,
-        'share_text': share_text,
-        'search_url': get_search_url_from_referer(request) or '',
-        'analytics_page_revision': doc.current_revision_id,
-        'analytics_en_slug': en_slug,
-        'content_experiment': rendering_params['experiment'],
-    }
-    response = render(request, 'wiki/document.html', context)
-    return _set_common_headers(doc, rendering_params['section'], response)
 
+@csrf_exempt
+@require_http_methods(['GET', 'HEAD', 'PUT'])
+@redirect_in_maintenance_mode(methods=['PUT'])
+@allow_CORS_GET
+@accepts_auth_key
+@process_document_path
+@newrelic.agent.function_trace()
+@ratelimit(key='user_or_ip', rate='100/m', block=True)
+def document_api(request, document_slug, document_locale):
+    """
+    View/modify the content of a wiki document, or create a new wiki document.
+    """
+    if request.method == 'PUT':
+        if not (request.authkey and request.user.is_authenticated()):
+            raise PermissionDenied
+        return _document_api_PUT(request, document_slug, document_locale)
+
+    try:
```

Comment:
I think `get_objects_or_404` can be used here.

Reply:
Good catch, thanks @safwanrahman !